internal/promotion/runner/builtin/metadata_setter.go

```diff
@@ -0,0 +1,147 @@
+package builtin
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/xeipuuv/gojsonschema"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	kargoapi "github.com/akuity/kargo/api/v1alpha1"
+	"github.com/akuity/kargo/pkg/promotion"
+	"github.com/akuity/kargo/pkg/x/promotion/runner/builtin"
+)
+
+// metadataSetter is an implementation of the promotion.StepRunner interface
+// that updates metadata on Stage or Freight resources.
+type metadataSetter struct {
+	kargoClient  client.Client
+	schemaLoader gojsonschema.JSONLoader
+}
+
+// newMetadataSetter returns an implementation of the promotion.StepRunner
+// interface that updates metadata on Stage or Freight resources.
+func newMetadataSetter(kargoClient client.Client) promotion.StepRunner {
+	r := &metadataSetter{
+		kargoClient: kargoClient,
+	}
+	r.schemaLoader = getConfigSchemaLoader(r.Name())
+	return r
+}
+
+// Name implements the promotion.StepRunner interface.
+func (s *metadataSetter) Name() string {
+	return "set-metadata"
+}
+
+// Run implements the promotion.StepRunner interface.
+func (s *metadataSetter) Run(
+	ctx context.Context,
+	stepCtx *promotion.StepContext,
+) (promotion.StepResult, error) {
+	cfg, err := s.convert(stepCtx.Config)
+	if err != nil {
+		return promotion.StepResult{
+			Status: kargoapi.PromotionStepStatusFailed,
+		}, &promotion.TerminalError{Err: err}
+	}
+	return s.run(ctx, stepCtx, cfg)
+}
+
+// convert validates the configuration against a JSON schema and converts it
+// into a builtin.SetMetadataConfig struct.
+func (s *metadataSetter) convert(cfg promotion.Config) (builtin.SetMetadataConfig, error) {
+	return validateAndConvert[builtin.SetMetadataConfig](s.schemaLoader, cfg, s.Name())
+}
+
+func (s *metadataSetter) run(
+	ctx context.Context,
+	_ *promotion.StepContext,
+	cfg builtin.SetMetadataConfig,
+) (promotion.StepResult, error) {
+
+	type updateGroup struct {
+		kind   string
+		values map[string]any
+	}
+
+	updatesByResource := make(map[string]*updateGroup)
+
+	for _, u := range cfg.Updates {
+		key := fmt.Sprintf("%s/%s", u.Kind, u.Name)
+		if _, exists := updatesByResource[key]; !exists {
+			updatesByResource[key] = &updateGroup{
+				kind:   string(u.Kind),
+				values: make(map[string]any),
+			}
+		}
+		for k, v := range u.Values {
+			updatesByResource[key].values[k] = v
+		}
+	}
```

Comment:
I appreciate you fixing the problem I previously raised of Stages or Freight being re-retrieved for every values change, but I think we may now have over-optimized.

Correct me if I am wrong, but compared to simply ranging over `cfg.Updates` directly starting on line +89, doesn't this optimization only provide any benefit in the unusual case that someone had unnecessarily done this:

```yaml
config:
  updates:
    - kind: Stage
      name: my-stage
      values:
        foo: bar
    - kind: Stage
      name: my-stage
      values:
        bat: baz
```

instead of this:

```yaml
config:
  updates:
    - kind: Stage
      name: my-stage
      values:
        foo: bar
        bat: baz
```

?